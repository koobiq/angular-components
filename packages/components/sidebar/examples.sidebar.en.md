### With splitter

<!-- example(sidebar-with-splitter) -->
