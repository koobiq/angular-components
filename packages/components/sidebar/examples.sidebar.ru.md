### Со сплиттером

<!-- example(sidebar-with-splitter) -->
