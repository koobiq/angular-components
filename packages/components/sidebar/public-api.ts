export * from './sidebar';
export * from './sidebar.module';
