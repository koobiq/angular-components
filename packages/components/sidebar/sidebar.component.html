<ng-container [ngSwitch]="internalState">
    <ng-container *ngSwitchCase="true">
        <ng-content select="[kbq-sidebar-opened]" />
    </ng-container>

    <ng-container *ngSwitchCase="false">
        <ng-content select="[kbq-sidebar-closed]" />
    </ng-container>
</ng-container>
