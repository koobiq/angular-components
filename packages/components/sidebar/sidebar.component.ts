import { DOCUMENT } from '@angular/common';
import {
    AfterContentInit,
    afterNextRender,
    ChangeDetectionStrategy,
    ChangeDetectorRef,
    Component,
    ContentChild,
    Directive,
    ElementRef,
    EventEmitter,
    inject,
    Input,
    NgZone,
    OnDestroy,
    Output,
    Renderer2,
    ViewEncapsulation
} from '@angular/core';
import { isControl, isInput, isLeftBracket, isRightBracket } from '@koobiq/cdk/keycodes';
import { kbqSidebarAnimations, KbqSidebarAnimationState } from './sidebar-animations';

export enum SidebarPositions {
    Left = 'left',
    Right = 'right'
}

interface KbqSidebarParams {
    openedStateMinWidth: string;
    openedStateWidth: string;
    openedStateMaxWidth: string;

    closedStateWidth: string;
}

@Directive({
    selector: '[kbq-sidebar-opened]',
    exportAs: 'kbqSidebarOpened',
    host: {
        class: 'kbq-sidebar-opened'
    }
})
export class KbqSidebarOpened {
    @Input() minWidth: string;
    @Input() width: string;
    @Input() maxWidth: string;
}

@Directive({
    selector: '[kbq-sidebar-closed]',
    exportAs: 'kbqSidebarClosed',
    host: {
        class: 'kbq-sidebar-closed'
    }
})
export class KbqSidebarClosed {
    @Input() width: string;
}

@Component({
    selector: 'kbq-sidebar',
    exportAs: 'kbqSidebar',
    templateUrl: 'sidebar.component.html',
    styleUrls: ['./sidebar.scss'],
    host: {
        class: 'kbq-sidebar',
        '[@state]': `{
            value: animationState,
            params: params
        }`,
        '(@state.start)': 'onAnimationStart()',
        '(@state.done)': 'onAnimationDone()'
    },
    animations: [kbqSidebarAnimations.sidebarState],
    encapsulation: ViewEncapsulation.None,
    changeDetection: ChangeDetectionStrategy.OnPush
})
export class KbqSidebar implements OnDestroy, AfterContentInit {
    /**
     * @docs-private
     */
    protected readonly document = inject<Document>(DOCUMENT);
    private readonly renderer = inject(Renderer2);
    private readonly changeDetectorRef = inject(ChangeDetectorRef);

    @Input()
    get opened(): boolean {
        return this._opened;
    }

    set opened(value: boolean) {
        if (this._opened) {
            this.saveWidth();
        }

        this._opened = value;
    }

    private _opened: boolean = true;

    @Input() position: SidebarPositions;

    /**
     * @docs-private
     */
    params: KbqSidebarParams = {
        openedStateWidth: 'inherit',
        openedStateMinWidth: 'inherit',
        openedStateMaxWidth: 'inherit',

        closedStateWidth: '32px'
    };

    @Output() readonly stateChanged: EventEmitter<boolean> = new EventEmitter<boolean>();

    /**
     * @docs-private
     */
    @ContentChild(KbqSidebarOpened, { static: false }) openedContent: KbqSidebarOpened;

    /**
     * @docs-private
     */
    @ContentChild(KbqSidebarClosed, { static: false }) closedContent: KbqSidebarClosed;

    /**
     * @docs-private
     */
    get animationState(): KbqSidebarAnimationState {
        return this._opened ? KbqSidebarAnimationState.Opened : KbqSidebarAnimationState.Closed;
    }

    /**
     * @docs-private
     */
    internalState: boolean = true;

    private unbindKeydownListener: ReturnType<Renderer2['listen']> | null = null;

    constructor(
        private ngZone: NgZone,
        private elementRef: ElementRef
    ) {
        afterNextRender(() => this.registerKeydownListener());
    }

    ngOnDestroy(): void {
        this.unRegisterKeydownListener();
    }

    ngAfterContentInit(): void {
        this.params = {
            openedStateWidth: this.openedContent.width || 'inherit',
            openedStateMinWidth: this.openedContent.minWidth || 'inherit',
            openedStateMaxWidth: this.openedContent.maxWidth || 'inherit',

            closedStateWidth: this.closedContent.width || '32px'
        };
    }

    toggle(): void {
        this.opened = !this.opened;
        this.changeDetectorRef.markForCheck();
    }

    /**
     * @docs-private
     */
    onAnimationStart() {
        if (this._opened) {
            this.internalState = this._opened;
        }
    }

    /**
     * @docs-private
     */
    onAnimationDone() {
        this.internalState = this._opened;

        this.stateChanged.emit(this._opened);
    }

    private registerKeydownListener(): void {
        this.ngZone.runOutsideAngular(() => {
            this.unbindKeydownListener ||= this.renderer.listen(this.document, 'keypress', (event) =>
                this.handleKeydown(event)
            );
        });
    }

    private unRegisterKeydownListener(): void {
        if (this.unbindKeydownListener) {
            this.unbindKeydownListener();
            this.unbindKeydownListener = null;
        }
    }

    private handleKeydown(event: KeyboardEvent): void {
        if (isControl(event) || isInput(event)) return;

        if (
            (this.position === SidebarPositions.Left && isLeftBracket(event)) ||
            (this.position === SidebarPositions.Right && isRightBracket(event))
        ) {
            this.toggle();
        }
    }

    private saveWidth() {
        this.params.openedStateWidth = `${this.elementRef.nativeElement.offsetWidth}px`;
    }
}
