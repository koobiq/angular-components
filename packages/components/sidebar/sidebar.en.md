Component designed to add collapsible side content.

<!-- example(sidebar-overview) -->

### Keyboard interaction

| <div style="min-width: 110px;">Key</div>       | Action                   |
| ---------------------------------------------- | ------------------------ |
| <span class="docs-hot-key-button">&#91;</span> | Open/close left Sidebar  |
| <span class="docs-hot-key-button">&#93;</span> | Open/close right Sidebar |
