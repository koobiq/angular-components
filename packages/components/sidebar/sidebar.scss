.kbq-sidebar {
    display: inline-block;
    height: 100%;
    overflow: hidden;

    .kbq-sidebar-opened,
    .kbq-sidebar-closed {
        height: 100%;
    }
}
