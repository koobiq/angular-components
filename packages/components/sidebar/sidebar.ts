import { animate, state, style, transition, trigger } from '@angular/animations';
import { DOCUMENT } from '@angular/common';
import {
    booleanAttribute,
    ChangeDetectionStrategy,
    Component,
    computed,
    contentChild,
    Directive,
    inject,
    input,
    NgZone,
    OnDestroy,
    OnInit,
    output,
    Renderer2,
    signal,
    ViewEncapsulation
} from '@angular/core';
import { outputFromObservable, outputToObservable, takeUntilDestroyed, toObservable } from '@angular/core/rxjs-interop';
import { isControl, isInput, isLeftBracket, isRightBracket } from '@koobiq/cdk/keycodes';
import { KbqAnimationDurations } from '@koobiq/components/core';
import { distinctUntilChanged } from 'rxjs/operators';

enum KbqSidebarAnimationState {
    Opened = 'opened',
    Closed = 'closed'
}

type KbqSidebarAnimationParams = Partial<{
    openedStateMinWidth: string;
    openedStateWidth: string;
    openedStateMaxWidth: string;
    closedStateWidth: string;
}>;

const KBQ_SIDEBAR_ANIMATION_PARAMS_DEFAULT: KbqSidebarAnimationParams = {
    openedStateMinWidth: 'inherit',
    openedStateWidth: 'inherit',
    openedStateMaxWidth: 'inherit',
    closedStateWidth: '32px'
};

const KBQ_SIDEBAR_ANIMATION = trigger('state', [
    state(
        KbqSidebarAnimationState.Opened,
        style({
            minWidth: '{{ openedStateMinWidth }}',
            width: '{{ openedStateWidth }}',
            maxWidth: '{{ openedStateMaxWidth }}'
        }),
        {
            params: {
                openedStateMinWidth: '',
                openedStateWidth: '',
                openedStateMaxWidth: ''
            } satisfies KbqSidebarAnimationParams
        }
    ),
    state(
        KbqSidebarAnimationState.Closed,
        style({
            minWidth: '{{ closedStateWidth }}',
            width: '{{ closedStateWidth }}',
            maxWidth: '{{ closedStateWidth }}'
        }),
        { params: { closedStateWidth: '' } satisfies KbqSidebarAnimationParams }
    ),
    transition(`${KbqSidebarAnimationState.Opened} => ${KbqSidebarAnimationState.Closed}`, [
        animate(KbqAnimationDurations.Entering)]),
    transition(`${KbqSidebarAnimationState.Closed} => ${KbqSidebarAnimationState.Opened}`, [
        animate(KbqAnimationDurations.Complex)])

]);

/**
 * Sidebar positions.
 */
export enum KbqSidebarPositions {
    Left = 'left',
    Right = 'right'
}

@Directive({
    standalone: true,
    selector: '[kbq-sidebar-opened],[kbqSidebarOpened]',
    exportAs: 'kbqSidebarOpened',
    host: {
        class: 'kbq-sidebar-opened'
    }
})
export class KbqSidebarOpened {
    /**
     * Min width of the sidebar when opened.
     */
    readonly minWidth = input<string>();
    /**
     * Width of the sidebar when opened.
     */
    readonly width = input<string>();
    /**
     * Max width of the sidebar when opened.
     */
    readonly maxWidth = input<string>();
}

@Directive({
    standalone: true,
    selector: '[kbq-sidebar-closed],[kbqSidebarClosed]',
    exportAs: 'kbqSidebarClosed',
    host: {
        class: 'kbq-sidebar-closed'
    }
})
export class KbqSidebarClosed {
    /**
     * Width of the sidebar when closed.
     */
    readonly width = input<string>();
}

@Component({
    standalone: true,
    selector: 'kbq-sidebar',
    exportAs: 'kbqSidebar',
    template: `
        @if (state()) {
            <ng-content select="[kbq-sidebar-opened],[kbqSidebarOpened]" />
        } @else {
            <ng-content select="[kbq-sidebar-closed],[kbqSidebarClosed]" />
        }
    `,
    styleUrl: './sidebar.scss',
    host: {
        class: 'kbq-sidebar',
        '[@state]': `{
            value: animationStateValue(),
            params: animationStateParams()
        }`,
        '(@state.start)': 'onAnimationStart()',
        '(@state.done)': 'onAnimationDone()'
    },
    animations: [KBQ_SIDEBAR_ANIMATION],
    encapsulation: ViewEncapsulation.None,
    changeDetection: ChangeDetectionStrategy.OnPush
})
export class KbqSidebar implements OnDestroy, OnInit {
    private readonly document = inject<Document>(DOCUMENT);
    private readonly ngZone = inject(NgZone);
    private readonly renderer = inject(Renderer2);

    private readonly openedContent = contentChild(KbqSidebarOpened);
    private readonly closedContent = contentChild(KbqSidebarClosed);

    /**
     * Whether the sidebar is opened or closed.
     */
    readonly opened = input(true, { transform: booleanAttribute });

    private readonly _opened = signal(this.opened());

    /**
     * Emits event when the sidebar opened state changes.
     * Also used for two-way binding.
     */
    readonly openedChange = output<boolean>();

    /**
     * Emits event when the sidebar opened state changes.
     *
     * @deprecated Will be removed in next major release, use `openedChange` instead.
     */
    readonly stateChanged = outputFromObservable(outputToObservable(this.openedChange).pipe(distinctUntilChanged()));

    /**
     * Sidebar position.
     */
    readonly position = input.required<KbqSidebarPositions>();

    /** @docs-private */
    protected readonly animationStateValue = computed<KbqSidebarAnimationState>(() => {
        return this._opened() ? KbqSidebarAnimationState.Opened : KbqSidebarAnimationState.Closed;
    });

    /** @docs-private */
    protected readonly animationStateParams = computed<KbqSidebarAnimationParams>(() => {
        const openedContent = this.openedContent();
        const closedContent = this.closedContent();

        return {
            openedStateMinWidth: openedContent?.minWidth() || KBQ_SIDEBAR_ANIMATION_PARAMS_DEFAULT.openedStateMinWidth,
            openedStateWidth: openedContent?.width() || KBQ_SIDEBAR_ANIMATION_PARAMS_DEFAULT.openedStateWidth,
            openedStateMaxWidth: openedContent?.maxWidth() || KBQ_SIDEBAR_ANIMATION_PARAMS_DEFAULT.openedStateMaxWidth,
            closedStateWidth: closedContent?.width() || KBQ_SIDEBAR_ANIMATION_PARAMS_DEFAULT.closedStateWidth
        };
    });

    /**
     * Internal opened state based on animations.
     *
     * @docs-private
     */
    protected readonly state = signal<boolean>(this._opened());

    private unbindKeydownListener: ReturnType<Renderer2['listen']> | null = null;

    constructor() {
        // @TODO: should be removed after migration to `linkedSignal`
        toObservable(this.opened)
            .pipe(takeUntilDestroyed())
            .subscribe((opened) => {
                this._opened.set(opened);
            });
    }

    ngOnInit(): void {
        this.registerKeydownListener();
    }

    ngOnDestroy(): void {
        this.unRegisterKeydownListener();
    }

    /**
     * Toggles the sidebar opened state.
     */
    toggle(): void {
        this._opened.update((opened) => !opened);
    }

    /** @docs-private */
    protected onAnimationStart(): void {
        const opened = this._opened();

        if (opened) {
            this.state.set(opened);
        }
    }

    /** @docs-private */
    protected onAnimationDone() {
        const opened = this._opened();

        this.state.set(opened);

        this.openedChange.emit(opened);
    }

    private registerKeydownListener(): void {
        this.ngZone.runOutsideAngular(() => {
            this.unbindKeydownListener ||= this.renderer.listen(this.document, 'keypress', (event) =>
                this.handleKeydown(event)
            );
        });
    }

    private unRegisterKeydownListener(): void {
        if (this.unbindKeydownListener) {
            this.unbindKeydownListener();
            this.unbindKeydownListener = null;
        }
    }

    private handleKeydown(event: KeyboardEvent): void {
        if (isControl(event) || isInput(event)) return;

        const position = this.position();

        if (
            (position === KbqSidebarPositions.Left && isLeftBracket(event)) ||
            (position === KbqSidebarPositions.Right && isRightBracket(event))
        ) {
            this.toggle();
        }
    }
}
