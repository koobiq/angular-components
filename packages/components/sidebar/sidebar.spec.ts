import { ChangeDetectionStrategy, Component, DebugElement, Provider, Type, viewChild } from '@angular/core';
import { ComponentFixture, fakeAsync, TestBed, tick } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { KbqSidebar, KbqSidebarPositions } from './sidebar';
import { KbqSidebarModule } from './sidebar.module';

const createComponent = <T>(component: Type<T>, providers: Provider[] = []): ComponentFixture<T> => {
    TestBed.configureTestingModule({
        imports: [component, NoopAnimationsModule],
        providers
    });
    const fixture = TestBed.createComponent<T>(component);

    fixture.autoDetectChanges();

    return fixture;
};

const getSidebarDebugElement = (debugElement: DebugElement): DebugElement => {
    return debugElement.query(By.directive(KbqSidebar));
};

@Component({
    standalone: true,
    selector: 'test-sidebar',
    imports: [KbqSidebarModule],
    template: `
        <kbq-sidebar [(opened)]="opened" [position]="position" (stateChanged)="onStateChanged($event)">
            <div kbqSidebarOpened>kbqSidebarOpened</div>
            <div kbqSidebarClosed>kbqSidebarClosed</div>
        </kbq-sidebar>
    `,
    changeDetection: ChangeDetectionStrategy.OnPush
})
class TestSidebar {
    readonly sidebar = viewChild.required(KbqSidebar);

    opened = true;

    position = KbqSidebarPositions.Left;

    readonly onStateChanged = jest.fn((_event: boolean) => {});
}

describe(KbqSidebar.name, () => {
    it('should match snapshot when opened', () => {
        const { debugElement } = createComponent(TestSidebar);

        expect(getSidebarDebugElement(debugElement)).toMatchSnapshot();
    });

    it('should match snapshot when closed', fakeAsync(() => {
        const { debugElement, componentInstance } = createComponent(TestSidebar);

        componentInstance.sidebar().toggle();
        tick();
        expect(getSidebarDebugElement(debugElement)).toMatchSnapshot();
    }));

    it('should emit two-way data binding by toggle method', fakeAsync(() => {
        const { componentInstance } = createComponent(TestSidebar);

        expect(componentInstance.opened).toBe(true);
        componentInstance.sidebar().toggle();
        tick();
        expect(componentInstance.opened).toBe(false);
    }));

    it('should emit stateChanged event by model change', fakeAsync(() => {
        const { componentInstance } = createComponent(TestSidebar);
        const spy = jest.spyOn(componentInstance, 'onStateChanged');

        expect(spy).toHaveBeenCalledTimes(0);
        componentInstance.opened = false;

        tick();
        expect(spy).toHaveBeenCalledTimes(1);
    }));

    it('should emit stateChanged event by toggle method', fakeAsync(() => {
        const { componentInstance } = createComponent(TestSidebar);
        const spy = jest.spyOn(componentInstance, 'onStateChanged');

        expect(spy).toHaveBeenCalledTimes(0);
        componentInstance.sidebar().toggle();

        tick();
        expect(spy).toHaveBeenCalledTimes(1);
    }));

    it('should emit stateChanged event with correct $event value by toggle method', fakeAsync(() => {
        const { componentInstance } = createComponent(TestSidebar);
        const spy = jest.spyOn(componentInstance, 'onStateChanged');

        componentInstance.sidebar().toggle();
        tick();
        expect(spy).toHaveBeenCalledWith(false);
    }));

    it('should close on `BracketLeft` key click', fakeAsync(() => {
        const { componentInstance } = createComponent(TestSidebar);

        expect(componentInstance.opened).toBe(true);
        expect(componentInstance.position).toBe(KbqSidebarPositions.Left);

        document.dispatchEvent(new KeyboardEvent('keypress', { code: 'BracketLeft' }));
        tick();
        expect(componentInstance.opened).toBe(false);
    }));

    it('should ignore on `BracketRight` key click', fakeAsync(() => {
        const { componentInstance } = createComponent(TestSidebar);

        expect(componentInstance.opened).toBe(true);
        expect(componentInstance.position).toBe(KbqSidebarPositions.Left);

        document.dispatchEvent(new KeyboardEvent('keypress', { code: 'BracketRight' }));
        tick();
        expect(componentInstance.opened).toBe(true);
    }));
});
