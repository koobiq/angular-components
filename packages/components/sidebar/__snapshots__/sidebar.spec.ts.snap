// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KbqSidebar should match snapshot when closed 1`] = `
DebugElement {
  "nativeNode": <kbq-sidebar
    class="kbq-sidebar ng-trigger ng-trigger-state"
    style="min-width: 32px; width: 32px; max-width: 32px;"
  >
    
    <div
      class="kbq-sidebar-closed"
      kbqsidebarclosed=""
      style=""
    >
      kbqSidebarClosed
    </div>
    
  </kbq-sidebar>,
}
`;

exports[`KbqSidebar should match snapshot when opened 1`] = `
DebugElement {
  "nativeNode": <kbq-sidebar
    class="kbq-sidebar ng-trigger ng-trigger-state"
    style="min-width: inherit; width: inherit; max-width: inherit;"
  >
    <div
      class="kbq-sidebar-opened"
      kbqsidebaropened=""
      style=""
    >
      kbqSidebarOpened
    </div>
    
    
  </kbq-sidebar>,
}
`;
